076-support-for-kafka-4.0.md

```diff
@@ -0,0 +1,200 @@
+
+## Motivation
+
+The move from Apache Kafka 3.x to 4.x and the related move from ZooKeeper-based Kafka clusters to KRaft is a major change for our users as well as for Strimzi itself.
+Having a good plan for how we plan to deal with it is important as it will provide clarity to Strimzi users, developers, and create a plan of the things that need to be done.
+It should also help our users to properly plan when to migrate from ZooKeeper to KRaft, as this is a major change that might require planning several months ahead of time.
+
+## Proposal
+
+Strimzi should approach adoption of Kafka 4.0 in the following way:
+
+1) When Apache Kafka 3.9.0 is released, we will adopt it in Strimzi as any other Apache Kafka release.
+   We will release it as part of a Strimzi release 0.4x which will support both Kafka 3.8.x and 3.9.x.
+   This version will support both ZooKeeper- and KRaft-based clusters.
+   This version will also add warnings to all users using ZooKeeper that it will be removed in the next Kafka release and that they should do the migration.
+
+2) Given the expected timeline of 3-4 months between Kafka 3.9 and 4.0, we will do another minor Strimzi release 0.4y.
+   This version might add support for some of the new features added to KRaft in Kafka 3.9.0.
+   0.4y will keep the support for 3.8 and 3.9 as well as support for both KRaft and ZooKeeper.
+   This version will be subject to the extended support described below in a dedicated section.
+   It will also keep the warnings introduced in previous Strimzi version.
+
+3) Once Strimzi 0.4y is released, we will proceed and remove ZooKeeper support from Strimzi.
+   This would include things such as:
+   * Deprecating the ZooKeeper related parts of the CRDs
+   * Removing the ZooKeeper reconciler and all other parts of the code relevant to ZooKeeper
+   * Remove the migration code
+   * Remove ZooKeeper related system tests
+
+   As upgrade to Strimzi 0.4z will not be possible without migration to KRaft, we will know for sure that the clusters run by Strimzi 0.4z had to be previously run by recent Strimzi versions.
+   That will also allow us to remove other legacy code.
+   For example the migration of Kafka clusters from StatefulSets to StrimziPodSets would not be needed anymore and would be removed.
+
+   Once Apache Kafka 4.0 is released, we will adopt it and do a Strimzi release 0.4z.
+   0.4z and all later Strimzi versions will support only KRaft based clusters.
+   Strimzi 0.4z will support the last two Kafka minor versions as usually - Kafka 3.9 and 4.0 - and will allow users to gradually upgrade as always.
+   But it will support only KRaft and users with ZooKeeper-based clusters will need to migrate to KRaft first within the Strimzi 0.4x and 0.4y versions and Kafka 3.9.
+
+4) The Strimzi releases after 0.4z will continue as usually with the 1-2 months release cycles and will continue to adopt the Kafka patch releases.
```

Comment:
```suggestion
4) The Strimzi releases after 0.4z will continue as usually with the roughly 2 months release cycles, and will continue to adopt the Kafka patch releases.
```

Reply:
Our goal always was to release every 1-2 months.